extensions/oplogPopulator/allocationStrategy/ImmutableConnector.js

```diff
@@ -0,0 +1,57 @@
+const AllocationStrategy = require('./AllocationStrategy');
+
+/**
+ * @class ImmutableConnector
+ *
+ * @classdesc ImmutableConnector is an allocation
+ * strategy that always requires a new connector to be
+ * created for each bucket.
+ */
+class ImmutableConnector extends AllocationStrategy {
+    /**
+     * Assign the bucket to a new connector. If a
+     * connector has retained the bucket, use it.
+     * If the bucket is already assigned to a connector,
+     * return the connector.
+     * @param {String} bucket bucket name
+     * @returns {Promise<Connector>} connector
+     */
+    async getOrAssignBucketConnector(bucket) {
+        let connector;
+        if (this._bucketsToConnectors.has(bucket)) {
+            return this._bucketsToConnectors.get(bucket);
+        } else if (this._retainedBuckets.has(bucket)) {
+            connector = this._retainedBuckets.get(bucket);
+        } else if (!connector) {
+            connector = await this._connectorsManager.addConnector();
+        }
+        this._bucketsToConnectors.set(bucket, connector);
+        await connector.addBucket(bucket);
+        return connector;
+    }
+
+    /**
+     * Delete a bucket from a connector, if it can be
+     * deleted
+     * @param {string} bucket bucket name
+     * @returns {Promise<Connector | null>} the modified connector returned if modified
+     */
+    async removeBucketFromConnector(bucket) {
```

Comment:
the only difference between both implementations is this line condition: `connector?.buckets.length === 1` vs `connector != nil`

--> this indicates the abstraction for this part of the interface is not at the right level

Having a simpler (non async) `removeBucketFromConnector(bucket, connector)` which simply returns true/false to indicate if the connector should be removed would do the job in a simpler way.